tokio/src/util/intrusive_double_linked_list.rs

```diff
@@ -0,0 +1,783 @@
+//! An intrusive double linked list of data
+
+#![allow(dead_code, unreachable_pub)]
+
+use core::{
+    marker::PhantomPinned,
+    ops::{Deref, DerefMut},
+    ptr::NonNull,
+};
+
+/// A node which carries data of type `T` and is stored in an intrusive list
+#[derive(Debug)]
+pub struct ListNode<T> {
+    /// The previous node in the list. `None` if there is no previous node.
+    prev: Option<NonNull<ListNode<T>>>,
+    /// The next node in the list. `None` if there is no previous node.
+    next: Option<NonNull<ListNode<T>>>,
+    /// The data which is associated to this list item
+    data: T,
+    /// Prevents `ListNode`s from being `Unpin`. They may never be moved, since
+    /// the list semantics require addresses to be stable.
+    _pin: PhantomPinned,
+}
+
+impl<T> ListNode<T> {
+    /// Creates a new node with the associated data
+    pub fn new(data: T) -> ListNode<T> {
+        ListNode::<T> {
+            prev: None,
+            next: None,
+            data,
+            _pin: PhantomPinned,
+        }
+    }
+}
+
+impl<T> Deref for ListNode<T> {
+    type Target = T;
+
+    fn deref(&self) -> &T {
+        &self.data
+    }
+}
+
+impl<T> DerefMut for ListNode<T> {
+    fn deref_mut(&mut self) -> &mut T {
+        &mut self.data
+    }
+}
+
+/// An intrusive linked list of nodes, where each node carries associated data
+/// of type `T`.
+#[derive(Debug)]
+pub struct LinkedList<T> {
+    head: Option<NonNull<ListNode<T>>>,
+    tail: Option<NonNull<ListNode<T>>>,
+}
+
+impl<T> LinkedList<T> {
+    /// Creates an empty linked list
+    pub fn new() -> Self {
+        LinkedList::<T> {
+            head: None,
+            tail: None,
+        }
+    }
+
+    /// Adds a node at the front of the linked list.
+    /// Safety: This function is only safe as long as `node` is guaranteed to
+    /// get removed from the list before it gets moved or dropped.
+    /// In addition to this `node` may not be added to another other list before
+    /// it is removed from the current one.
+    pub unsafe fn add_front(&mut self, node: &mut ListNode<T>) {
```

Comment:
The intrusive list @carllerche proposed in #2210 requires nodes to be pinned in order to be inserted into the list: 
https://github.com/tokio-rs/tokio/blob/18906454273d77cc0faa2c736b9e0079eee3f583/tokio/src/util/linked_list.rs#L58

I have a strong preference for this API level, as it helps us enforce one of the two safety invariants documented in the above comment (that the inserted node's address will remain stable and will not drop while it's in the list).